epidose/device/update.sh

```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
+#
+# Updates software
+#
+# Copyright 2020 Stefanos Georgiou
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+set -e
+
+export APP_NAME=update_epidose
+
+# Pick up utility functions relative to the script's source code
+UTIL="$(dirname "$0")/util.sh"
+
+# Source common functionality (logging, WiFi)
+# shellcheck source=epidose/device/util.sh
+SERVER_URL=none . "$UTIL"
+
+cd /home/epidose/epidose
+log "Updating software"
+make fast-install &
```

Comment:
Why is it executing asynchronously? It can create a race condition.

Reply:
In case I remote the `&` the I get the following error:

```bash
make: *** wait: No child processes.  Stop.
make: *** Waiting for unfinished jobs....
make: *** wait: No child processes.  Stop.
```

Any idea of how I can do it differently?

Reply:
No idea. What happens when you run make by hand?  Are you sure that the asynchronous run doesn't just hide the errors?  Try redirecting its output to a temporary file to see it.

Reply:
1. When I execute the `make fast-install` by hand, the program copies new source code files in the `/opt/venvs/epidose/bin/` and restarts the episode daemons.

2. I am not sure, but I can see that all epidose daemons are starting again, after the execution of the `make fast-install &` (the uptime from `supervisorctl status` shows that), but the program crashes when I execute it as `make fast-install`.
Also, I have tried to add a log message that is pulled and printed as expected.

3. I will try that too and let you know.

Reply:
I have tried to redirect the output (`make fast-install & 2>&1 tmp`) but the tmp file was empty after the update operation.
Is this the correct way to redirect the output?

Reply:
I have tried it like this: `make fast-install > /home/epidose/udpate_with_ampersand.txt 2>&1 &`
and this is the only output that I see inside the temp file:
```bash
cp -r epidose dp3t /opt/venvs/epidose/lib/python3.7/site-packages/
cp epidose/device/shutdown_epidose.sh \
        epidose/device/upload_seeds_d.sh \
	epidose/device/update_filter_d.sh \
	epidose/device/util.sh \
	epidose/device/wps_scanner_d.sh /opt/venvs/epidose/bin/
supervisorctl restart epidose:*
epidose:upload_seeds: stopped
epidose:beacon_rx: stopped
epidose:wps_scanner: stopped
epidose:watchdog: ERROR (spawn error)
epidose:beacon_tx: started
epidose:beacon_rx: started
epidose:update_filter: started
epidose:wps_scanner: started
epidose:upload_seeds: started
```

and this is the output of `update_filter_d`:
```bash
2020-11-12 16:56:27 Sleeping for 21597 s
2020-11-12 16:56:48 Killed during sleep; obtaining a new filter
2020-11-12 16:56:48 Obtaining new filter from http://ha-server:5010
2020-11-12 16:56:48 Acquiring WiFi
2020-11-12 16:56:48 Turn on WiFi
2020-11-12 16:56:48 Temporary stop beacon transmissions
2020-11-12 16:56:50 Acquired WiFi
2020-11-12 16:56:50 New filter obtained: 105 bytes
2020-11-12 16:56:50 Checking for updates
2020-11-12 16:56:51 Update is needed
From https://github.com/stefanos1316/epidose
 * branch            HEAD       -> FETCH_HEAD
2020-11-12 16:56:52 Starting up
2020-11-12 16:56:52 Updating software
2020-11-12 16:56:55 Starting up
2020-11-12 16:56:55 Filter's age (5 s) is current; no update required
2020-11-12 16:56:55 Sleeping for 21595 s
```

Here I have actually killed the sleep time in order to force the program to continue,
but before doing that, I have changed the watchdog_d's heartbeat rate.
As a result, it pulled the new software and build the package again.

Reply:
Can you please try with `make -j1`?

Reply:
`make -j1` without the ampersand you mean?

Reply:
Yes.

Reply:
I have tried after pulling the new changes to do `make -j1 fast-install` and  I get the following message:
```bash
From https://github.com/eellak/epidose
 * branch            production -> FETCH_HEAD
2020-11-12 18:20:55 Starting up
2020-11-12 18:20:55 Updating software
make: *** [Makefile:28: fast-install] Terminated
```

Reply:
It seems that `make` terminates with an error.  You don't see the error, because I think that stderr is redirected to an error file.  Redirect the standard error of make (without `&`) to a file to see it.  The line `epidose:watchdog: ERROR (spawn error)` provides a hint on what is going on.  The `& just masked the error.

Reply:
I have tried it as `make -j1 fast-install > /home/epidose/update_errors.txt 2>&1` and I get the following output:
```bash
cp -r epidose dp3t /opt/venvs/epidose/lib/python3.7/site-packages/
cp epidose/device/shutdown_epidose.sh \
        epidose/device/upload_seeds_d.sh \
	epidose/device/update_filter_d.sh \
	epidose/device/util.sh \
	epidose/device/wps_scanner_d.sh /opt/venvs/epidose/bin/
supervisorctl restart epidose:*
make: *** wait: No child processes.  Stop.
make: *** Waiting for unfinished jobs....
make: *** wait: No child processes.  Stop.
```

Reply:
I have tried to add the `make fast-install` command in the `update-software.sh` to see where the script fails.
It seems that it fails the moment it executes the `supervisorctl restart epidose:*` because it stops and tries to start the process again. However, one of these processes is the one that calls the`supervisorctl`.

Reply:
Here is one solution that I have in mind. We can have the `update_software` as a running daemon that is going to be in `sleep infinity` state until a `git pull` is executed (which is going to kill the background infinity sleep process). Then, the `update_software` daemon can restart all the processes apart from itself. A drawback of this solution, however, it that we cannot make changes to the `update_software` daemon if needed.

Reply:
Good catch!  I think the way to handle this issue is to provide to the update script (not the daemon) enough to do one of at least two things:
1. install and restart all processes apart from the update daemon
2. install and reboot the computer, so that the new update daemon will run

Reply:
1) The current commit does, to my opinion, what you are suggesting at your first point. I have also created a separate rule in the Makefile named as `remote-update` which restarts all the daemons but `update_software_d`.
This is what you have in mind?

2) But honestly, I think we should still implement a rule and if there is a change in the `update_software.sh`, then we should reboot the device, no?